src/main/java/au/org/emii/ncdfgenerator/NcdfEncoderBuilder.java

```diff
@@ -0,0 +1,1476 @@
+
+	public void addValueToBuffer( Object value );
+
+	public String getName(); // change class name to IVariableEncoder and this to just getName()
+
+}
+
+
+interface IDimension extends IAddValue
+{
+	public void define( NetcdfFileWriteable writer) ;
+
+	public Dimension getDimension( ) ; // horrible to expose this...
+										// can't the caller create the dimension?
+	//
+	public int getLength();
+
+	public void addValueToBuffer( Object value );
+
+	public String getName();
+
+}
+
+
+class DimensionImpl implements IDimension
+{
+
+	public DimensionImpl( String name )
+	{
+		this.name = name; // required to encode dimension
+		this.size = 0;
+	}
+
+	final String name;
+	int size;
+	Dimension dimension;
+
+
+	public Dimension getDimension( )  // bad naming
+	{
+		// throw if not defined...
+		return dimension;
+	}
+
+	public int getLength()
+	{
+		return size;
+	}
+
+	public void define( NetcdfFileWriteable writer)
+	{
+		dimension = writer.addDimension( name, size );
+	}
+
+	public void addValueToBuffer( Object value )
+	{
+		++size;
+	}
+
+	public String getName()
+	{
+		return name ;
+
+	}
+
+}
+
+
+class VariableEncoder implements IVariableEncoder
+{
+	public VariableEncoder(
+		String variableName,
+		ArrayList< IDimension> dimensions,
+		IValueEncoder encodeValue,
+		Map<String, String> attributes
+	) {
+		this.variableName = variableName;
+		this.encodeValue = encodeValue;
+		this.attributes = attributes;
+		this.dimensions = dimensions;
+		this.buffer = new ArrayList<Object>( );
+	}
+
+	final String variableName;
+	final IValueEncoder			encodeValue;
+	final Map<String, String>	attributes;
+	final ArrayList<IDimension>	dimensions; // change name childDimensions
+	final ArrayList<Object>		buffer;
+
+
+	/*	we can also record the table, or index of table here if we want
+			to incorporate into the strategy.
+		eg. we can compre with xml to decide what to do.
+	*/
+	public void addValueToBuffer( Object value )
+	{
+		// perhaps delegate to strategy...
+		buffer.add( value );
+	}
+
+	public void define( NetcdfFileWriteable writer )
+	{
+		// write dims and attributes
+
+		// make sure children are defined already
+		List<Dimension> d = new ArrayList<Dimension>();
+		for( IDimension dimension: dimensions)
+		{
+			d.add( dimension.getDimension() );
+		}
+
+		writer.addVariable(variableName, encodeValue.targetType(), d );
+
+		for( Map.Entry< String, String> entry : attributes.entrySet()) {
+			writer.addVariableAttribute( variableName, entry.getKey(), entry.getValue()/*.toString()*/ );
+		}
+	}
+
+
+	public void writeValues( ArrayList<IDimension> dims, int dimIndex, int acc, Array A  )
+		throws NcdfGeneratorException
+	{
+		if( dimIndex < dims.size() )
+		{
+			Dimension dim = dims.get( dimIndex ).getDimension();
+			for( int i = 0; i < dim.getLength(); i++ )
+			{
+				writeValues( dims, dimIndex + 1, acc + i, A );
+			}
+		}
+		else
+		{
+			// System.out.println( "dimIndex " + "  acc " + acc  + "  buffer " + buffer.get( acc ) );
```

Comment:
I've removed a lot of these, but will leave a couple of high level ones while further dev is required, before finally pulling them out.